RR - use exception mappers on auth failure exceptions for proactive auth

michalvavrik opened the pull request:
fixes: #29896

This is proposal to enable RESTEasy Reactive exception mappers to handle `AuthenticationCompletionException`, `AuthenticationFailedException`, `AuthenticationRedirectException`, `ForbiddenException` when `quarkus.http.auth.proactive=true`. I suggest we weight the arguments and decide whether the change is needed in some form or the PR should be closed.

Pros:
- user can use exception mappers when `quarkus.http.auth.proactive=false`
- it's JAX-RS way that user asked for in the past
- HTTP path matching policy may lead to 403 that will be customizable via exception mappers (after this PR, not down to `proactive=true`), currently it's only customizable via failure handlers
- by accident, above-mentioned already works for RESTEasy Classic since 2.14 (however we can regress it) and classic/reactive behavior should be same

Cons:
- IMO in reported issues, users seemed complacent about using `proactive=false` (once Sergey suggested it)
- custom exception mapper that catches `AuthenticationFailedException` thrown by OIDC is unlikely to send challenge
- currenlty users can customize response via failure handlers after the default auth failure handler has send challenge and failed event (iff auth mechanism didn't end response), therefore it's safer

sberyozkin commented:
@michalvavrik Sorry for the delay.
Can you please add a test showing that fault handlers can still be used with the proactive auth enabled if preferred unless we already have the existing test ? I'm not sure why you remove the doc section explaining how to use the fault handlers.
Thanks

michalvavrik commented:
@sberyozkin sure, I'll think about it and may add tests next week. I removed the docs section as it says with `proactive=true` the only way to customize response is to use failure handler, which would no more be true. My issue with user defined failure handler when `proactive=true` is that with RESTEasy Reactive/Classic they would need negative priority and I don't want to advice that to users. I think primary way for RESTEasy Reactive/Classic should be exception mappers and failure handlers are available for special cases or for non-JAX-RS users (Reactive Routes, pure lambdas?).

michalvavrik commented:
Maybe there is another way, let me think about it, I'm currently on sick leave.

sberyozkin commented:
@michalvavrik Sorry, take care, no rush at all, lets talk in a few days (I'd consider keeping the doc section but adjusting the advice only when/how to use it)

sberyozkin commented:
@michalvavrik Hey, so the only question on my part here, is the fault route path still working with this PR (can you link to the test please), and if yes, then my only change request is to keep the info about using fault routes in the docs (with some minor updates - it won't longer be the only option)

michalvavrik commented:
> @michalvavrik Hey, so the only question on my part here, is the fault route path still working with this PR (can you link to the test please), and if yes, then my only change request is to keep the info about using fault routes in the docs (with some minor updates - it won't longer be the only option)

Simply priority adjustment will make failure handlers working again and then I will return docs section & add all relevant tests. I'll do that.

michalvavrik commented:
I've added an examle and caution note to docs regarding auth mechanism challenge. Together existing test coverage and added tests covers both exception mappers and failure handlers in RESTEasy Reactive and Classic.

sberyozkin commented on docs/src/main/asciidoc/security-built-in-authentication-support-concept.adoc:
```diff
@@ -187,7 +187,40 @@ public class HelloService {
+}
+----
+
+CAUTION: By default, authentication mechanisms like `io.quarkus.oidc.runtime.OidcAuthenticationMechanism` needs to handle Quarkus Security authentication exceptions themselves in order redirect request and set correct response status.
```
```suggestion
CAUTION: Some HTTP authentication mechanisms such as `io.quarkus.oidc.runtime.OidcAuthenticationMechanism` need to handle authentication exceptions themselves in order to create a correct redirect URL, status code, cookies, etc.
```

sberyozkin replied:
Sorry, just realized OIDC handles bearer tokens as well which is just 401, so we need to be specific it is the code flow exceptions which should not be customized

sberyozkin commented on docs/src/main/asciidoc/security-built-in-authentication-support-concept.adoc:
```diff
@@ -187,7 +187,40 @@ public class HelloService {
+----
+
+CAUTION: By default, authentication mechanisms like `io.quarkus.oidc.runtime.OidcAuthenticationMechanism` needs to handle Quarkus Security authentication exceptions themselves in order redirect request and set correct response status.
+For that reason, built-in exception mappers that handles authentication failures challenge authentication mechanisms.
```
```suggestion
For that reason, using custom exception mappers to customize authentication exceptions thrown by mechanisms such as `OIDC` is not recommended.
```

sberyozkin commented on docs/src/main/asciidoc/security-built-in-authentication-support-concept.adoc:
```diff
@@ -187,7 +187,40 @@ public class HelloService {
+
+CAUTION: By default, authentication mechanisms like `io.quarkus.oidc.runtime.OidcAuthenticationMechanism` needs to handle Quarkus Security authentication exceptions themselves in order redirect request and set correct response status.
+For that reason, built-in exception mappers that handles authentication failures challenge authentication mechanisms.
+See xref:security-customization.adoc#dealing-with-more-than-one-http-auth-mechanisms[Dealing with more than one HttpAuthenticationMechanism] section for more details on authentication mechanism challenge and more specifically {quarkus-blob-url}/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/exceptionmappers/SecurityExceptio ...
```
```suggestion
In such cases, a safer way to customize authentication exceptions is to make sure the proactive authentication is not disabled and use Vert.x HTTP route failure handlers, as events come to the handler with the correct response status and headers.
```

sberyozkin replied:
I thought the last sentence was too complex, with a link to the source is likely to be too technical kind of info for the docs, so propose just to replace it with a simple advice

michalvavrik replied:
I think suggested way is clearer thanks.

sberyozkin commented on docs/src/main/asciidoc/security-built-in-authentication-support-concept.adoc:
```diff
@@ -187,7 +187,40 @@ public class HelloService {
+For that reason, built-in exception mappers that handles authentication failures challenge authentication mechanisms.
+See xref:security-customization.adoc#dealing-with-more-than-one-http-auth-mechanisms[Dealing with more than one HttpAuthenticationMechanism] section for more details on authentication mechanism challenge and more specifically {quarkus-blob-url}/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/exceptionmappers/SecurityExceptio ...
+
+Perhaps safer way with enabled Proactive Authentication is to provide a failure handler, as event comes to the handler with correct response status and headers,
```
It can be now removed if the above suggestion is merged

sberyozkin commented on docs/src/main/asciidoc/security-built-in-authentication-support-concept.adoc:
```diff
@@ -187,7 +187,40 @@ public class HelloService {
+See xref:security-customization.adoc#dealing-with-more-than-one-http-auth-mechanisms[Dealing with more than one HttpAuthenticationMechanism] section for more details on authentication mechanism challenge and more specifically {quarkus-blob-url}/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/exceptionmappers/SecurityExceptio ...
+
+Perhaps safer way with enabled Proactive Authentication is to provide a failure handler, as event comes to the handler with correct response status and headers,
+therefore only thing that needs to be done is to customize response like this:
```
```suggestion
therefore the only thing that needs to be done is to customize the response like this:
```

sberyozkin commented:
@michalvavrik Great work, thanks, left a few minor doc suggestions only

michalvavrik committed:
RR - use exception mappers on auth failure exceptions for proactive auth

#29896

michalvavrik commented:
Docs looks better with your comments, thank you.
